miqcli/api.py

```diff
@@ -63,16 +68,79 @@ def connect(self, settings):
         :param settings: ManageIQ settings.
         :type settings: dict
         """
+        authfile = os.path.join(os.environ["HOME"], AUTHDIR)
+        # attempt a connection with a passed token
+        if "token" in settings and settings["token"]:
+            succ_auth, exception = self.miq_auth(settings["token"])
+            if not succ_auth:
+                raise RuntimeError("Unable to auth with passed token: "
+                                   "{0}".format(exception))
+        # check for an auth token file & validate
+        if os.path.isfile(authfile):
+            with file(authfile) as f:
+                settings["token"] = f.read().strip()
+            succ_auth, exception = self.miq_auth(settings["token"])
+            if not succ_auth:
+                if "username" in settings and settings["username"] and \
+                        "password" in settings and settings["password"]:
+                    self.get_token()
+                else:
+                    raise RuntimeError("Please provide valid "
+                                       "credentials to connect")
+                succ_auth, exception = self.miq_auth(settings["token"])
+                if not succ_auth:
+                    raise RuntimeError("Unable to authenticate "
+                                       "{0}".format(exception))
+        # authenticate and create auth token file if it does not exist
+        else:
+            if not os.path.exists(os.path.dirname(authfile)):
+                os.makedirs(os.path.dirname(authfile))
+
+            if "username" in settings and settings["username"] and \
+                    "password" in settings and settings["password"]:
+                self.get_token()
+            else:
+                raise RuntimeError("Provide valid credentials to connect")
+            succ_auth, exception = self.miq_auth(settings["token"])
+            if not succ_auth:
+                raise RuntimeError("Authentication failed: "
+                                   "{0}".format(exception))
+
+    def get_token(self):
+        """
+        method to get a token if username/password is set
+        """
+        auth_endpoint = self.settings["url"] + "/auth"
```

Comment:
Looks like the **settings** attribute is unresolved. This attribute is only in the scope of the Collection class, which is a child class of the parent class ClientAPI. If we want the ClientAPI class to have an attribute for the MIQ configuration settings. We should look into passing the settings to the ClientAPI class. Then the Collection class would still have a settings attribute.

```python
-    def __init__(self):
-        """Constructor."""
+    def __init__(self, settings):
+        """Constructor.
+
+        :param settings: MIQ settings
+        :type settings: dict
+        """
+        self.settings = settings
         self._client = None
```

Then within the miqcli/collections/__init__.py we could remove the settings attribute from there.

```python
-    _settings = None
-
     def __init__(self, settings):
         """
         :param settings: MIQ settings
         :type settings: dict
         """
-        super(Collection, self).__init__()
-        self._settings = settings
-
-    @property
-    def settings(self):
-        """
-        :return: dict of settings
-        """
-        return self._settings
+        super(Collection, self).__init__(settings)
```

Those code snippets were not tested, just put as an example.